[Connector API] Implement update scheduling action

jedrazb opened the pull request:
## Changes

- Implement `PUT _connector/{connector_id}/scheduling` endpoint
- This endpoint expects body with a new connector scheduling object, it overwrites the existing `scheduling` field in a given connector doc

## Verification
- Add unit tests for index service
- Add BWC tests
- Add YML tests

Also, verified new endpoint manually.

jedrazb committed:
[Connector API] Implement update scheduling action

jedrazb committed:
Improve comments, fix typos

jedrazb committed:
Apply formatting✨

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:Enterprise Search)

jedrazb committed:
Merge branch 'main' into connectors-api-update-scheduling

jedrazb committed:
Fix typo

carlosdelest commented on rest-api-spec/src/main/resources/rest-api-spec/api/connector.update_scheduling.json:
```diff
@@ -0,0 +1,39 @@
+    "url": {
+      "paths": [
+        {
+          "path": "/_connector/{connector_id}/scheduling",
```
I'd _say_ that actions done on a resource in ES tend to have underscore:

```suggestion
          "path": "/_connector/{connector_id}/_scheduling",
```

I think that's more in line with what we have. WDYT @kderusso ? (Or anyone that can keep me honest here)

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -166,6 +169,36 @@ public void onFailure(Exception e) {
+        try {
+            String connectorId = request.getConnectorId();
+            final UpdateRequest updateRequest = new UpdateRequest(CONNECTOR_INDEX_NAME, connectorId).doc(
+                new IndexRequest(CONNECTOR_INDEX_NAME).opType(DocWriteRequest.OpType.INDEX)
```
I'm not sure if this applies as it's part of the update request, but I think the proper operation should be update. 

```suggestion
                new IndexRequest(CONNECTOR_INDEX_NAME).opType(DocWriteRequest.OpType.UPDATE)
```

Let's try this with implementing the 404 test mentioned above and see if it works correctly 👍

jedrazb replied:
That's what I thought as well - but the IndexRequest doesn't allow to `OpType.UPDATE` ([IndexRequest source code](https://github.com/elastic/elasticsearch/blob/714aa680f11514ee981682785423bc4d3bb7d383/server/src/main/java/org/elasticsearch/action/index/IndexRequest.java#L216)) - My understanding is that we are creating a new scheduling object (overwriting a field in a parent connector doc)

carlosdelest replied:
Thanks for confirming! 👍

carlosdelest reviewed (approved):
Good job! You got the hang of it 🚀 

A few comments about testing and paths that should be considered. They can be done in a follow up PR.

jedrazb commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/331_connector_update_scheduling.yml:
```diff
@@ -0,0 +1,60 @@
+        body:
+          incremental:
+            enabled: false
+            interval: 3 0 0 * * ?
```
Yeah that makes sense, let me add the test in this PR :)

jedrazb commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/331_connector_update_scheduling.yml:
```diff
@@ -0,0 +1,60 @@
+            interval: 3 0 0 * * ?
+
+  - match:
+      status: 400
```
Yes will add 👍

jedrazb committed:
Add more yaml tests, change API path to _scheduling

jedrazb commented:
@carlosdelest I added more tests testing for 404s and 400 with malformed cron expression. Also updated the API path to`_scheduling`. Thank you for quick feedback! :)
